From ec97aa1aeebd4bcacaa6bebbba75f1809bb7e35a Mon Sep 17 00:00:00 2001
From: Kajetan Puchalski <kajetan.puchalski@tuta.io>
Date: Thu, 8 Oct 2020 23:29:59 +0100
Subject: [PATCH] Toggle keys patch

Allows toggling all dwm keybinds on and off.
When off, only the special toggle keybind remains active.
Useful for instance in games that use the same modifier keys as dwm.
---
 config.def.h |  2 ++
 dwm.c        | 19 +++++++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/config.def.h b/config.def.h
index 1c0b587..288b52f 100644
--- a/config.def.h
+++ b/config.def.h
@@ -5,6 +5,7 @@ static const unsigned int borderpx  = 1;        /* border pixel of windows */
 static const unsigned int snap      = 32;       /* snap pixel */
 static const int showbar            = 1;        /* 0 means no bar */
 static const int topbar             = 1;        /* 0 means bottom bar */
+static int enablekeys               = 1;
 static const char *fonts[]          = { "monospace:size=10" };
 static const char dmenufont[]       = "monospace:size=10";
 static const char col_gray1[]       = "#222222";
@@ -61,6 +62,7 @@ static const char *termcmd[]  = { "st", NULL };
 
 static Key keys[] = {
 	/* modifier                     key        function        argument */
+	{ MODKEY,                       XK_z,      togglekeys,     {0} },
 	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
 	{ MODKEY|ShiftMask,             XK_Return, spawn,          {.v = termcmd } },
 	{ MODKEY,                       XK_b,      togglebar,      {0} },
diff --git a/dwm.c b/dwm.c
index 664c527..53ccfc6 100644
--- a/dwm.c
+++ b/dwm.c
@@ -234,6 +234,7 @@ static int xerror(Display *dpy, XErrorEvent *ee);
 static int xerrordummy(Display *dpy, XErrorEvent *ee);
 static int xerrorstart(Display *dpy, XErrorEvent *ee);
 static void zoom(const Arg *arg);
+static void togglekeys(const Arg *arg);
 
 /* variables */
 static const char broken[] = "broken";
@@ -992,6 +993,18 @@ keypress(XEvent *e)
 
 	ev = &e->xkey;
 	keysym = XKeycodeToKeysym(dpy, (KeyCode)ev->keycode, 0);
+
+	if (!enablekeys) {
+		for (i = 0; i < LENGTH(keys); i++)
+			if (keys[i].func == togglekeys
+				&& keysym == keys[i].keysym
+				&& CLEANMASK(keys[i].mod) == CLEANMASK(ev->state)) {
+				enablekeys = 1;
+				return;
+			}
+		return;
+	}
+
 	for (i = 0; i < LENGTH(keys); i++)
 		if (keysym == keys[i].keysym
 		&& CLEANMASK(keys[i].mod) == CLEANMASK(ev->state)
@@ -999,6 +1012,12 @@ keypress(XEvent *e)
 			keys[i].func(&(keys[i].arg));
 }
 
+void
+togglekeys(const Arg* arg)
+{
+	enablekeys ^= 1;
+}
+
 void
 killclient(const Arg *arg)
 {
-- 
2.28.0

